numpy/linalg/linalg.py

```diff
@@ -1538,11 +1538,34 @@ def cond(x, p=None):
 
     """
     x = asarray(x)  # in case we have a matrix
-    if p is None:
+    if p is None or p == 2 or p == -2:
         s = svd(x, compute_uv=False)
-        return s[..., 0]/s[..., -1]
+        with errstate(all='ignore'):
+            if p == -2:
+                r = s[..., -1] / s[..., 0]
+            else:
+                r = s[..., 0] / s[..., -1]
     else:
-        return norm(x, p, axis=(-2, -1)) * norm(inv(x), p, axis=(-2, -1))
+        # Call inv(x) ignoring errors: the result array will be
+        # contain nans in the entries where inversion failed.
+        _assertRankAtLeast2(x)
+        _assertNdSquareness(x)
+        t, result_t = _commonType(x)
+        signature = 'D->D' if isComplexType(t) else 'd->d'
+        with errstate(all='ignore'):
+            invx = _umath_linalg.inv(x, signature=signature)
+            r = norm(x, p, axis=(-2, -1)) * norm(invx, p, axis=(-2, -1))
+        r = r.astype(result_t, copy=False)
+
+    if r.ndim > 0:
+        r[~isfinite(r)] = Inf
+    elif not isfinite(r):
+        if isinstance(r, ndarray):
+            r[()] = Inf
+        else:
+            r = type(r)(Inf)
```

Comment:
How about:

```python
r = asarray(r)
r[~isfinite(r)] = Inf
return r[()]
```

Reply:
Ok, changed. Also fixed handling of nans in inputs.